File: src/native/schemas/routing.ts
import { z } from "zod";
import { toolIdSchema } from "./tool";

const safeCountSchema = z
  .number()
  .int()
  .nonnegative()
  .max(Number.MAX_SAFE_INTEGER);

export const routingProviderSchema = z
  .object({
    id: z.string().min(1).max(256),
    name: z.string().min(1).max(256),
    priority: z.number().int().positive().max(512).nullable(),
    current: z.boolean(),
    healthy: z.boolean(),
    consecutiveFailures: z.number().int().nonnegative().max(4_294_967_295),
  })
  .strict();

export const routingTargetSchema = z
  .object({
    tool: toolIdSchema,
    takeoverEnabled: z.boolean(),
    autoFailoverEnabled: z.boolean(),
    currentProvider: routingProviderSchema.nullable(),
    queue: z.array(routingProviderSchema).max(512),
    available: z.array(routingProviderSchema).max(512),
  })
  .strict();

export const routingOverviewSchema = z
  .object({
    running: z.boolean(),
    address: z.string().min(1).max(255).nullable(),
    port: z.number().int().positive().max(65_535).nullable(),
    activeConnections: safeCountSchema,
    totalRequests: safeCountSchema,
    successRequests: safeCountSchema,
    failedRequests: safeCountSchema,
    failoverCount: safeCountSchema,
    targets: z.array(routingTargetSchema).length(4),
  })
  .strict()
  .superRefine((overview, context) => {
    if (
      overview.running !== (overview.address !== null && overview.port !== null)
    ) {
      context.addIssue({
        code: "custom",
        message: "running state and endpoint must agree",
      });
    }
    if (new Set(overview.targets.map((target) => target.tool)).size !== 4) {
      context.addIssue({
        code: "custom",
        message: "routing targets must be unique",
      });
    }
  });

export type RoutingProvider = z.infer<typeof routingProviderSchema>;
export type RoutingTarget = z.infer<typeof routingTargetSchema>;
export type RoutingOverview = z.infer<typeof routingOverviewSchema>;

File: src-tauri/src/domain/routing.rs
//! Advanced Mode local-routing projection (ADR-0007).
//!
//! The inherited proxy engine contains credentials, raw request errors and live
//! configuration backups. None of those belong on the product wire. This file
//! deliberately exposes only aggregate counters, provider identity and health.

use serde::{Deserialize, Serialize};

use super::ToolId;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoutingProvider {
    pub id: String,
    pub name: String,
    /// One-based queue priority. `None` means the provider is available but is
    /// not currently part of the failover queue.
    pub priority: Option<u32>,
    pub current: bool,
    pub healthy: bool,
    pub consecutive_failures: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoutingTarget {
    pub tool: ToolId,
    pub takeover_enabled: bool,
    pub auto_failover_enabled: bool,
    pub current_provider: Option<RoutingProvider>,
    pub queue: Vec<RoutingProvider>,
    pub available: Vec<RoutingProvider>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoutingOverview {
    pub running: bool,
    /// Present only while the local server is running.
    pub address: Option<String>,
    pub port: Option<u16>,
    pub active_connections: u64,
    pub total_requests: u64,
    pub success_requests: u64,
    pub failed_requests: u64,
    pub failover_count: u64,
    pub targets: Vec<RoutingTarget>,
}

#[cfg(test)]
mod tests {
    use super::{RoutingOverview, RoutingProvider, RoutingTarget};
    use crate::domain::ToolId;

    #[test]
    fn wire_projection_contains_no_secrets_paths_notes_or_raw_errors() {
        let overview = RoutingOverview {
            running: true,
            address: Some("127.0.0.1".to_string()),
            port: Some(15_721),
            active_connections: 1,
            total_requests: 4,
            success_requests: 3,
            failed_requests: 1,
            failover_count: 1,
            targets: vec![RoutingTarget {
                tool: ToolId::ClaudeCode,
                takeover_enabled: true,
                auto_failover_enabled: true,
                current_provider: Some(RoutingProvider {
                    id: "provider-a".to_string(),
                    name: "Provider A".to_string(),
                    priority: Some(1),
                    current: true,
                    healthy: true,
                    consecutive_failures: 0,
                }),
                queue: vec![],
                available: vec![],
            }],
        };

        let json = serde_json::to_value(overview).expect("serialize routing overview");
        let object = json.as_object().expect("overview object");
        assert_eq!(object.get("activeConnections"), Some(&serde_json::json!(1)));
        let encoded = json.to_string().to_ascii_lowercase();
        for forbidden in [
            "token",
            "secret",
            "api_key",
            "apikey",
            "settingsconfig",
            "provider_notes",
            "providernotes",
            "last_error",
            "lasterror",
            "backup",
            "path",
        ] {
            assert!(!encoded.contains(forbidden), "wire leaked {forbidden}");
        }
    }
}
